tools/xmlschema.py

```diff
@@ -0,0 +1,462 @@
+
+        required = "required" if self.element.attrib["required"] == "1" else "optional"
+        el.set("use", required)
+
+        if required == "optional" and "default" in self.element.attrib:
+            el.set("default", self.element.attrib["default"])
+
+        return namespaces, el
+
+
+class ComplexType:
+    def __init__(self, element: ElementTree.Element, name: str = None):
+        self.element = element
+        self.name = name
+
+    def to_subtree(self):
+        namespaces = list()
+        elements = list()
+        attributes = list()
+
+        for attribute in self.element.findall("attribute"):
+            child_ns, child_el = Attribute(attribute).to_subtree()
+            attributes.append(child_el)
+            namespaces += child_ns
+
+        for child in self.element.findall("element"):
+            child_ns, child_el = Element(child).to_subtree()
+            elements.append(child_el)
+            namespaces += child_ns
+
+        for child in self.element.findall("include"):
+            child_ns, child_el = Include(child).to_subtree()
+            elements.append(child_el)
+            namespaces += child_ns
+
+        el = ElementTree.Element(_to_qname("xs:complexType"))
+
+        if self.name:
+            el.set("name", self.name)
+
+        docs = self.element.find("description")
+        if docs is not None:
+            doc_namespaces, doc_el = Description(docs).to_subtree()
+            namespaces.extend(doc_namespaces)
+            if doc_el:
+                el.append(doc_el)
+
+        if elements and "type" in self.element.attrib:
+            raise NotImplementedError(
+                "Cant handle sub-elements for an element declaring a type."
+            )
+        elif "type" in self.element.attrib:
+            el_type = _to_simple_type(self.element.attrib["type"])
+            extension = ElementTree.Element(_to_qname("xs:extension"))
+            extension.set("base", el_type)
+            if el_type.startswith("types"):
+                namespaces.append("types")
+            extension.extend(attributes)
+            simple_content = ElementTree.Element(_to_qname("xs:simpleContent"))
+            simple_content.append(extension)
+            el.append(simple_content)
+        elif elements:
+            # drop depreciated elements
+            elements = [el for el in elements if not el.attrib["maxOccurs"] == "0"]
+            unbounded_elements = [el for el in elements if el.attrib["maxOccurs"] == "unbounded"]
+
+            if not unbounded_elements:
+                seq_container = ElementTree.Element(_to_qname("xs:all"))
+                seq_container.extend(elements)
+                el.append(seq_container)
+                el.extend(attributes)
+            else:
+                # XSD 1.1 allows maxOccurs="unbound" within xs:all, so 
+                # we could drop this entire else block if we could upgrade
+                # to XSD 1.1. (xmllint only supports XSD 1.0)
```

Comment:
This isn't an issue with XSD but rather an issue with sdformat. XSD v1.0 (the schema version used here) is limited in that it doesn't allow us to represent the true capabilities of SDF. 

In SDF, child elements may appear 0, 1, or infinitely many times depending on the element and - this is what kills XSD 1.0 - may do so in any order. In XSD you may choose _exactly_ one element from a set of elements (`xs:choice`), you may specify a block in which elements may occur in any order and with a specified (and finite) min/max count for each element (`xs:all`), or you may specify a block in which elements must occur in the given order but may do so the given (potentially infinite) number of times. Neither of these choices fits.

The way the ruby generator solved this is by aggregating all valid children into a `xs:choice` element which may repeat 0 to inf many times. This works for elements with infinite counts but doesn't enforce maxOccurs for elements that may only occur once. If we, thus, validate SDF files with the resulting schema, it may produce false negatives by saying that a SDF is valid even though a unique element occurs twice.

The way the commented out code solves this is by grouping elements into a group with finite count (`<=1`) and a group with infinite count. It then inserts the `xs:choice` block discussed above (for infinite-count elements only) between every two fine-count elements. It then does this for every permutation of finite-count elements (because in SDF elements may occur in any order, but we need to validate that unique elements don't reoccur). This will result in XSD that doesn't produce above false negatives, but it comes at the cost of _massive_ schemata (GBs in magnitude). 

So for now I just do the same here as the ruby script (accept that the result produces false negatives), because we can't change how SDF works, and we also can't bump the XSD version without prior feedback from one of the maintainers.

In XSD 1.1 this is less of a problem because above mentioned `xs:all` block allows elements with infinite counts. This means that we can simply use an `xs:all` block instead of a `xs:choice` block and be happy.

Reply:
Thanks for the detailed explanation. :+1: 
My comment was just about `xmllint only supports XSD 1.0` which is probably a reason why we can't simply use XSD 1.1 as there is no tool for linting. An option might be to only use XSD 1.1 if there is no other choice (= you need XSD 1.1 to represent it properly) and otherwise stick with XSD 1.0 - I don't know if mixing the two is even supported.